File: src/Functional/OCaml/patterns/observer.ml
type observer = string -> unit
type subscription = string * observer

let subscribe key observer subscribers =
  if List.exists (fun (existing_key, _) -> String.equal existing_key key) subscribers then
    (false, subscribers)
  else (true, (key, observer) :: subscribers)

let unsubscribe key subscribers =
  let remaining =
    List.filter (fun (existing_key, _) -> not (String.equal existing_key key)) subscribers
  in
  (List.length remaining <> List.length subscribers, remaining)

let publish event subscribers =
  List.iter (fun (_, notify) -> notify event) (List.rev subscribers)

let () =
  let audit_events = ref [] in
  let dashboard_events = ref [] in
  let audit event = audit_events := event :: !audit_events in
  let dashboard event = dashboard_events := event :: !dashboard_events in
  let added_audit, subscribers = subscribe "audit" audit [] in
  let added_dashboard, subscribers = subscribe "dashboard" dashboard subscribers in
  let duplicate_dashboard, subscribers = subscribe "dashboard" dashboard subscribers in
  assert added_audit;
  assert added_dashboard;
  assert (not duplicate_dashboard);
  publish "ready" subscribers;
  assert (!audit_events = [ "ready" ]);
  assert (!dashboard_events = [ "ready" ]);
  let removed_dashboard, subscribers = unsubscribe "dashboard" subscribers in
  let removed_dashboard_again, subscribers = unsubscribe "dashboard" subscribers in
  assert removed_dashboard;
  assert (not removed_dashboard_again);
  publish "done" subscribers;
  assert (!audit_events = [ "done"; "ready" ]);
  assert (!dashboard_events = [ "ready" ])

File: src/Functional/Haskell/Observer.hs
module Observer (Observer, Subscription, subscribe, unsubscribe, publish, examplePasses) where

type Observer a b = a -> b
type Subscription a b = (String, Observer a b)

subscribe :: String -> Observer a b -> [Subscription a b] -> [Subscription a b]
subscribe key observer subscriptions
  | any ((== key) . fst) subscriptions = subscriptions
  | otherwise = subscriptions ++ [(key, observer)]

unsubscribe :: String -> [Subscription a b] -> [Subscription a b]
unsubscribe key = filter ((/= key) . fst)

publish :: [Subscription a b] -> a -> [b]
publish subscriptions value = map (($ value) . snd) subscriptions

examplePasses :: Bool
examplePasses =
  let audit value = "audit:" ++ show value
      dashboard value = "dashboard:" ++ show value
      initial = subscribe "audit" audit []
      both = subscribe "dashboard" dashboard initial
      duplicateRejected = subscribe "audit" audit both
      afterUnsubscribe = unsubscribe "audit" duplicateRejected
   in publish duplicateRejected (42 :: Int) == ["audit:42", "dashboard:42"]
        && length duplicateRejected == 2
        && publish afterUnsubscribe 43 == ["dashboard:43"]
